File: app/src/main/java/com/rubyhuntersky/peregrine/model/AllAccounts.java
package com.rubyhuntersky.peregrine.model;

import com.rubyhuntersky.peregrine.ui.UiHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

/**
 * @author wehjin
 * @since 11/16/15.
 */

public class AllAccounts {
    public static final String JSONKEY_ACCOUNTS = "accounts";
    public static final String JSONKEY_ARRIVAL_DATE = "arrivalDate";

    public List<EtradeAccount> accounts;
    public Date arrivalDate;

    public AllAccounts(List<EtradeAccount> accounts, Date arrivalDate) {
        this.accounts = accounts;
        this.arrivalDate = arrivalDate;
    }

    public AllAccounts(String jsonAccountList) throws JSONException {
        accounts = new ArrayList<>();
        final JSONObject jsonObject = new JSONObject(jsonAccountList);
        final JSONArray jsonArray = jsonObject.getJSONArray(JSONKEY_ACCOUNTS);
        for (int i = 0; i < jsonArray.length(); i++) {
            accounts.add(new EtradeAccount(jsonArray.getJSONObject(i)));
        }
        arrivalDate = new Date(jsonObject.getLong(JSONKEY_ARRIVAL_DATE));
    }

    public CharSequence getRelativeArrivalTime() {
        return UiHelper.getRelativeTimeString(arrivalDate.getTime());
    }

    public JSONObject toJSONObject() throws JSONException {
        final JSONObject accountListJsonObject = new JSONObject();
        accountListJsonObject.put(JSONKEY_ARRIVAL_DATE, arrivalDate.getTime());

        final JSONArray jsonArray = new JSONArray();
        for (EtradeAccount account : accounts) {
            try {
                jsonArray.put(account.toJSONObject());
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        accountListJsonObject.put(JSONKEY_ACCOUNTS, jsonArray);
        return accountListJsonObject;
    }
}

File: app/src/main/java/com/rubyhuntersky/peregrine/model/AllAccounts.kt
package com.rubyhuntersky.peregrine.model

import com.rubyhuntersky.peregrine.ui.UiHelper
import com.rubyhuntersky.peregrine.utility.getDate
import com.rubyhuntersky.peregrine.utility.getJSONObjectList
import com.rubyhuntersky.peregrine.utility.putDate
import com.rubyhuntersky.peregrine.utility.putJSONObjectList
import org.json.JSONObject
import java.util.*

/**
 * @author wehjin
 * *
 * @since 11/16/15.
 */

class AllAccounts(val accounts: List<EtradeAccount>, val arrivalDate: Date) {

    val relativeArrivalTime: CharSequence
        get() = UiHelper.getRelativeTimeString(arrivalDate.time)

    companion object {
        val JSONKEY_ACCOUNTS = "accounts"
        val JSONKEY_ARRIVAL_DATE = "arrivalDate"

        @JvmStatic
        fun fromJson(string: String): AllAccounts {
            val jsonObject = JSONObject(string)
            return AllAccounts(
                    accounts = jsonObject.getJSONObjectList(JSONKEY_ACCOUNTS).map(::EtradeAccount),
                    arrivalDate = jsonObject.getDate(JSONKEY_ARRIVAL_DATE)
            )
        }

        @JvmStatic
        fun toJson(allAccounts: AllAccounts): String {
            return allAccounts.toJSONObject().toString()
        }

        fun AllAccounts.toJSONObject(): JSONObject {
            val jsonObject = JSONObject()
            with(jsonObject) {
                putDate(JSONKEY_ARRIVAL_DATE, arrivalDate)
                putJSONObjectList(JSONKEY_ACCOUNTS, accounts.map { it.toJSONObject() })
            }
            return jsonObject
        }
    }
}
